README.md
## GitHub auth provider for Autohost

### config

```javascript
{
  "auth": {
    "sessionMessages": false, // true to add login failure messages to session
    "loginEndpoint": "/auth/login", // this is the public endpoint for your login page
    "authEndpoint": "/auth/github", // this is the route that will kickoff the authentication against GitHub
    "github": {
        "organization": "ORG-NAME", // optionally add to require org membership
        "clientId": "YOUR_CLIENT_ID_HERE",
        "clientSecret": "YOUR_CLIENT_SECRET_HERE",
        "callbackUrl": "http://localhost:3000/auth/github/callback" // this is the route the GitHub will request after authentication
    }
  }
}
```


CHANGELOG.md
## 1.0.0-beta-1

- Added option to require organization membership

### 1.0.0-beta

### Initial Release




spec/setup.js
global.when = require( "when" );
global.chai = require( "chai" );
chai.use( require( "chai-as-promised" ) );
global.should = chai.should();
global.expect = chai.expect;
global._ = require( "lodash" );
global.fs = require( "fs" );
global.sinon = require( "sinon" );
/*eslint-disable */
var sap = require( "sinon-as-promised" );
chai.use( require( "sinon-chai" ) );
chai.use( require( "chai-passport-strategy" ) );
/*eslint-enable */
global.proxyquire = require( "proxyquire" ).noPreserveCache().noCallThru();
global.nock = require( "nock" );

chai.Assertion.addMethod( "returnError", function( message ) {
	var obj = this._obj;
	if ( !obj.then ) {
		obj = global.when.resolve( obj );
	}
	var self = this;
	return obj.then( function( err ) {
		new chai.Assertion( err ).to.be.instanceof( Error );
		return self.assert(
			err.message === message,
			"expected error message to be \"#{exp}\" but got \"#{act}\"",
			message,
			err.message
		);
	} );
} );

chai.Assertion.addMethod( "partiallyEql", function( partial ) {
	var obj = this._obj;
	if ( !obj.then ) {
		obj = global.when.resolve( obj );
	}
	var self = this;
	return obj.then( function( actual ) {
		var diffs = deepCompare( partial, actual );
		return self.assert(
			diffs.length === 0,
			diffs.join( "\n\t" )
		);
	} );
} );

function consoleLulz( original ) {
	var _log = console[ original ];
	console[ original ] = function() {
		if ( typeof arguments[ 0 ] === "string" && /^(\S|\t|\s)/.test( arguments[ 0 ] ) ) {
			_log.apply( console, arguments );
		} else {
			return; // swallow this message
		}
	};
}
consoleLulz( "log" );
consoleLulz( "warn" );
consoleLulz( "error" );

function deepCompare( a, b, k ) {
	var diffs = [];
	if ( b === undefined ) {
		diffs.push( "expected " + k + " to equal " + a + " but was undefined " );
	} else if ( _.isObject( a ) || _.isArray( a ) ) {
		_.each( a, function( v, c ) {
			var key = k ? [ k, c ].join( "." ) : c;
			diffs = diffs.concat( deepCompare( a[ c ], b[ c ], key ) );
		} );
	} else {
		var equal = a == b; // jshint ignore:line
		if ( !equal ) {
			diffs.push( "expected " + k + " to equal " + a + " but got " + b );
		}
	}
	return diffs;
}


spec/index.spec.js
require( "./setup" );
var GitHubAuthProvider = require( "../src" );

describe( "Autohost GitHub auth provider", function () {
	var provider;
	var defaultConfig = {
		auth: {
			github: {
				organization: "UrMom",
				clientId: "kds73osf9298fsj2f78j",
				clientSecret: "abc123ssshhhhhhhh"
			},
			roles: {
				client: [ "TeamOne" ],
				agent: [ "TeamTwo" ],
				admin: [ "TeamThree" ]
			},
			defaultRoles: [ "readonly" ]
		}
	};

	beforeEach(function () {
		provider = new GitHubAuthProvider( defaultConfig );
	});

	it( "should contain a GitHub auth strategy instance", function () {
		var GitHubStrategy = require( "passport-github2" ).Strategy;
		provider.strategies.length.should.equal( 1 );
		provider.strategies[ 0 ].should.be.an.instanceof( GitHubStrategy );
	} );
} );


spec/githubApi.spec.js
require( "./setup" );
var GitHubApi = require( "../src/githubApi" );

describe( "GitHubApi wrapper", function () {
	describe( "validateUserOrg function", function () {
		it( "should return true if the request returns a status of 204", function ( done ) {
			var path = "/orgs/ExampleInc/members/somedeveloperusername";
			nock( "https://api.github.com/" )
				.get( path )
				.reply( 204 );

			GitHubApi.validateUserOrg( "testToken", "somedeveloperusername", "ExampleInc", function ( err, isOrgMember ) {
				should.not.exist( err );
				isOrgMember.should.equal( true );
				done();
			} );
		} );
		it( "should return false if the request returns a status of anything other than 204", function ( done ) {
			var path = "/orgs/ExampleInc/members/somedeveloperusername";
			nock( "https://api.github.com/" )
				.get( path )
				.reply( 404 );

			GitHubApi.validateUserOrg( "testToken", "somedeveloperusername", "ExampleInc", function ( err, isOrgMember ) {
				should.not.exist( err );
				isOrgMember.should.equal( false );
				done();
			} );
		} );
		it( "should set the appropriate headers on the request", function ( done ) {
			var path = "/orgs/ExampleInc/members/somedeveloperusername";
			nock( "https://api.github.com/" )
				.matchHeader( "User-Agent", /^autohost-github-auth\// )
				.matchHeader( "Authorization", "token testToken" )
				.get( path )
				.reply( 204 );

			GitHubApi.validateUserOrg( "testToken", "somedeveloperusername", "ExampleInc", function ( err, isOrgMember ) {
				should.not.exist( err );
				isOrgMember.should.equal( true );
				done();
			} );
		} );
		it( "should fail when the incorrect token is sent with the request", function ( done ) {
			var path = "/orgs/ExampleInc/members/somedeveloperusername";
			nock( "https://api.github.com/" )
				.get( path )
				.reply( 401 );

			GitHubApi.validateUserOrg( "wrongToken", "somedeveloperusername", "ExampleInc", function ( err, isOrgMember ) {
				should.not.exist( err );
				done();
			} );
		} );
	} );
	describe( "loadUserTeams function", function () {
		var response = require( "./mocks/teams-response" );
		it( "should return an array of teams for the given user", function ( done ) {
			var path = "/user/teams";
			nock( "https://api.github.com" )
				.get( path )
				.reply( 200, response );

			GitHubApi.loadUserTeams( "testToken", function ( err, teams ) {
				should.not.exist( err );
				teams.should.eql( response );
				done();
			} );
		} );
		it( "should set the appropriate headers on the request", function ( done ) {
			var path = "/user/teams";
			nock( "https://api.github.com/" )
				.matchHeader( "User-Agent", /^autohost-github-auth\// )
				.matchHeader( "Authorization", "token testToken" )
				.get( path )
				.reply( 200 );

			GitHubApi.loadUserTeams( "testToken", function ( err, teams ) {
				should.not.exist( err );
				done();
			} );
		} );
		it( "should fail when the incorrect token is sent with the request", function ( done ) {
			var path = "/user/teams";
			nock( "https://api.github.com/" )
				.get( path )
				.reply( 401 );

			GitHubApi.loadUserTeams( "wrongToken", function ( err, body ) {
				err.should.be.an.instanceof( Error );
				done();
			} );
		} );
	} );
} );


src/index.js
var when = require( 'when' );
var githubStrategy = require( './githubStrategy' );

module.exports = function( config ) {
	var githubAuth;
	var useSession = !( config == undefined ? false : config.noSession );

	function authenticate( req, res, next ) {
		if ( req.path.indexOf( config.auth.authEndpoint ) === 0 ) {
			return githubAuth( req, res, next );
		}

		if ( req.path.indexOf( config.auth.loginEndpoint ) === 0 ) {
			return res.status( 500 ).send( "Login endpoint should not be authenticated" );
		}

		res.redirect( config.auth.loginEndpoint );
	}

	function deserializeUser( user, done ) {
		done( null, user);
	}

	function serializeUser( user, done ) {
		done( null, user );
	}

	var wrapper = {
		authenticate: authenticate,
		checkPermission: function () { return when( true ); },
		deserializeUser: deserializeUser,
		getActionRoles: function () { return []; },
		getUserRoles: function (user) {
			return when( user.roles );
		},
		hasUsers: function () { return when( true ); },
		initPassport: function( passport ) {
			githubAuth = passport.authenticate( 'github', { scope: ['user:email', 'read:org'], failureMessage: !!config.auth.sessionMessages, failureRedirect: config.auth.loginEndpoint, session: useSession } );
		},
		serializeUser: serializeUser,
		strategies: [
			githubStrategy( config.auth )
		],
		updateActions: function () { return when( true ); }
	};

	return wrapper;
};


src/githubStrategy.js
var _ = require( 'lodash' );
var GitHubStrategy = require( 'passport-github2' ).Strategy;
var githubApi = require( './githubApi' );

function loadUserRoles( accessToken, teamRoleMap, defaultRoles, org, done ) {
	if( !teamRoleMap ) {
		return done( null, defaultRoles );
	}

	githubApi.loadUserTeams( accessToken, function( err, teams ) {
		if( err ) {
			return done( err );
		}

		org = org.toLowerCase();
		var roles = teams.reduce( function( result, team ) {
			if( team.organization.login.toLowerCase() !== org ) {
				return result;
			}

			return result.concat( teamRoleMap[ team.name.toLowerCase() ] || [] );
		}, [] );

		done( null, _.uniq( roles.concat( defaultRoles ) ) );
	} );
}

module.exports = function ( config ) {
	var org = config.github.organization;
	var roles = config.roles;
	var defaultRoles = config.defaultRoles || [];

	// Create map of team->roles for faster lookup
	var teamRoleMap = roles && _.reduce( roles, function( map, teams, role ) {
		teams.forEach( function( team ) {
			team = team.toLowerCase();

			if( !map[ team ] ) {
				map[ team ] = [];
			}

			map[ team ].push( role );
		} );

		return map;
	}, {} );

	var github = new GitHubStrategy( {
		clientID: config.github.clientId,
		clientSecret: config.github.clientSecret,
		callbackUrl: config.github.callbackUrl
	}, function( accessToken, refreshToken, profile, done ) {
		if( !org ) {
			return process.nextTick( function() {
				done( null, profile );
			});
		}

		githubApi.validateUserOrg( accessToken, profile.username, org, function( err, isOrgMember ) {
			if( err ) {
				return done( err );
			}

			if( !isOrgMember ) {
				return done( null, false, { message: "User is not a member of the " + org + " organization." } );
			}

			loadUserRoles( accessToken, teamRoleMap, defaultRoles, org, function( err, roles ) {
				if( err ) {
					return done( err );
				}

				profile.roles = roles;

				done( null, profile );
			} );
		} );
	} );

	return github;
};


src/githubApi.js
var pkg = require( "../package.json" );
var url = require( "url" );
var request = require( "request" );

function fetch( token, path, done ) {
	request( {
		method: "GET",
		url: url.resolve( "https://api.github.com", path),
		headers: {
			"User-Agent": "autohost-github-auth/" + pkg.version,
			"Authorization": "token " + token
		},
		json: true
	}, function( err, res, body ) {
		if ( err ) {
			return done( err );
		}

		if ( res.statusCode < 200 || res.statusCode >= 300 ) {
			return done( new Error( "HTTP error. Status: " + res.statusCode ), res, body );
		}

		return done( null, res, body );
	} );
};

module.exports = {
	validateUserOrg: function ( accessToken, username, org, done ) {
		var path = "/orgs/" + org + "/members/" + username;
		return fetch( accessToken, path, function ( err, res, body ) {
			if ( res ) {
				return done( null, res.statusCode === 204 );
			}

			done( err, false );
		} );
	},

	loadUserTeams: function ( accessToken, done ) {
		var path = "/user/teams";
		fetch( accessToken, path, function ( err, res, body ) {
			done( err, body );
		} );
	}
};
